More granular User API Key scopes, and plugin support

davidtaylorhq opened the pull request:
This PR includes three commits, which I intend to "rebase and merge", so that they remain separate. Please check each commit message for details. 

The overall aim here is to allow core/plugins to create a user api key, scoped to a specific route with specific parameters.

danielwaterworth commented:
This is looking good, but it feels like the `RouteMatcher` is getting constructed at the wrong time. At the moment you're having to pass in `allowed_param_values` to the match method, but I think it would be clearer if the `RouteMatcher` had all of the data necessary to match against an arbitrary request. You could either achieve this by either allowing a `RouteMatcher` to be refined after construction, i.e. `RouteMatcher.new(...).with_allowed_param_values(...)` or not putting `RouteMatchers` into the constant.

davidtaylorhq commented:
> it feels like the RouteMatcher is getting constructed at the wrong time. At the moment you're having to pass in allowed_param_values to the match method

All the parameters which `RouteMatcher` is initialized are constant. The "allowed param values" vary at runtime based on user-provided configuration. So they do need to be separated somehow, but we can certainly try and improve the ergonomics.

> You could achieve this by either allowing a RouteMatcher to be refined after construction, i.e. RouteMatcher.new(...).with_allowed_param_values(...)

I would be a little worried about making this thread-safe. `with_allowed_param_values` would have to return a fresh RouteMatcher to avoid polluting the unrefined RouteMatcher. Is that what you had in mind?

> not putting RouteMatchers into the constant.

Yeah I wasn't too sure about this... it feels a little weird. But one major advantage is that it means missing/extra initialisation parameters are detected at boot, rather than at runtime.

danielwaterworth commented:
> I would be a little worried about making this thread-safe. with_allowed_param_values would have to return a fresh RouteMatcher to avoid polluting the unrefined RouteMatcher. Is that what you had in mind?

Yes, a fresh `RouteMatcher`.

I bring this up, because the name `RouteMatcher` implies to me that it has everything that it needs to match a route.

romanrizzi commented on lib/auth/default_current_user_provider.rb:
```diff
@@ -336,7 +337,7 @@ def lookup_api_user(api_key_value, request)
 
       params = @env['action_dispatch.request.parameters']
-      unless api_key.request_allowed?(request, params)
+      unless api_key.request_allowed?(@env)
```
I think we don't need to assign the `params` variable anymore if we pass the env directly.

romanrizzi commented on lib/plugin/instance.rb:
```diff
@@ -797,15 +797,33 @@ def add_api_key_scope(resource, action)
+        methods: methods,
+        actions: actions,
+        formats: formats
+      ), self)
```
Do we want to drop the old parameters at some point? If so, maybe we should a deprecation notice?

davidtaylorhq commented:
Thanks @romanrizzi and @danielwaterworth - have made some improvements based on your comments.

davidtaylorhq commented:
Fixed the failing specs, and updated the plugin API to automatically prefix scope names with the plugin name to prevent clashes (thanks for the suggestion @danielwaterworth)
